arangod/RestServer/SoftShutdownFeature.cpp

```diff
@@ -0,0 +1,234 @@
+
+#include "ApplicationFeatures/ApplicationServer.h"
+#include "ApplicationFeatures/ShutdownFeature.h"
+#include "FeaturePhases/AgencyFeaturePhase.h"
+#include "GeneralServer/GeneralServerFeature.h"
+#include "Logger/Logger.h"
+#include "Logger/LoggerFeature.h"
+#include "Pregel/PregelFeature.h"
+#include "RestServer/ConsoleFeature.h"
+#include "RestServer/DatabaseFeature.h"
+#include "RestServer/MetricsFeature.h"
+#include "RestServer/ScriptFeature.h"
+#include "RestServer/SoftShutdownFeature.h"
+#include "Scheduler/SchedulerFeature.h"
+#include "Scheduler/SupervisedScheduler.h"
+#include "Transaction/Manager.h"
+#include "Transaction/ManagerFeature.h"
+#include "Utils/CursorRepository.h"
+#include "VocBase/vocbase.h"
+
+using namespace arangodb::options;
+
+namespace {
+
+bool queueShutdownChecker(std::mutex& mutex,
+    arangodb::Scheduler::WorkHandle& workItem,
+    std::function<void(bool)>& checkFunc) {
+
+  arangodb::Scheduler* scheduler = arangodb::SchedulerFeature::SCHEDULER;
+  bool queued = false;
+  std::lock_guard<std::mutex> guard(mutex);
+  std::tie(queued, workItem) =
+      scheduler->queueDelay(arangodb::RequestLane::CLUSTER_INTERNAL,
+                            std::chrono::seconds(2),
+                            checkFunc);
+  return queued;
+}
+
+}
+
+namespace arangodb {
+
+SoftShutdownFeature::SoftShutdownFeature(application_features::ApplicationServer& server)
+    : ApplicationFeature(server, "SoftShutdown") {
+  setOptional(true);
+  startsAfter<application_features::AgencyFeaturePhase>();
+  startsAfter<ShutdownFeature>();
+  startsAfter<ConsoleFeature>();
+  startsAfter<ScriptFeature>();
+
+  // We do not yet know if we are a coordinator, so just in case,
+  // create a SoftShutdownTracker, it will not hurt if it is not used:
+  _softShutdownTracker = std::make_shared<SoftShutdownTracker>(server);
+}
+
+void SoftShutdownFeature::beginShutdown() {
+  _softShutdownTracker->cancelChecker();
+}
+
+void SoftShutdownTracker::cancelChecker() {
+  if (_softShutdownOngoing.load(std::memory_order_relaxed)) {
+    // This is called when an actual shutdown happens. We then want to
+    // delete the WorkItem of the soft shutdown checker, such that the
+    // Scheduler does not have any cron jobs any more:
+    _workItem.reset();
+  }
+}
+
+SoftShutdownTracker::SoftShutdownTracker(
+    application_features::ApplicationServer& server)
+  : _server(server), _softShutdownOngoing(false) {
+  _checkFunc = [this](bool cancelled) {
+    if (_server.isStopping()) {
+      return;   // already stopping, do nothing, and in particular
+                // let's not schedule ourselves again!
+    }
+    if (!this->check()) {   // Initiates shutdown if counters are 0
+      // Rearm ourselves:
+      if (!::queueShutdownChecker(this->_workItemMutex, this->_workItem,
+                                  this->_checkFunc)) {
+        // If this does not work, shut down right away:
+        this->initiateActualShutdown();
+      }
```

Comment:
This looked unintuitive to me, although after looking up the functions, I think it works.
My stylistic issues with it: `check()` sounds like it would be side-effects free. It is also declared `const`. But it can trigger the shutdown, which was unexpected to me.
I would prefer if `check()` was really side-effects-free and that the caller would trigger any require side-effects instead.

Reply:
Let's rename the method to `checkAndShutdownIfAllClear`, will do.

Reply:
Done.